### PR TITLE
[libc++] <experimental/simd> Add copy functions for class simd/simd_mask

### PR DESCRIPTION
**Pull request by joy2myself**

### CODE REVIEW

**Comment by joy2myself** [comment]
@philnik777 @ldionne Hi! I think we may need to discuss how to further reduce the compilation time of SIMD tests. With the current test methods and coverage, the compilation time of this patch may be very long, and the compilation time of subsequent patches such as SIMD binary operators may be even longer. Do you have any suggestions?

**Comment by philnik777** [comment]
Do you know why the test is taking so long?

**Comment by joy2myself** [comment]
> Do you know why the test is taking so long?

@philnik777 Louis has mentioned that the main impact on compile time is the instantiation of classes.

In previous tests, we generally traversed all **vectorizable data types** and all **simd ABIs** to instantiate various `simd`/`simd_mask` classes for testing. In this case, there would be a total of **20 (data types) x 21 (ABIs)** times instantiation of a class. However, some interfaces are template functions (such as the copy function in the current PR), which require additional traversal of the types of the template parameter `U`, that is, a total of **20 (T types) x 20 (U types) x 21 (ABIs)** instantiations. In addition, there are 3 different **alignment flags** involved in the copy function, so this test has a total of **20 (T types) x 20 (U types) x 21 (ABIs) x 3 (alignment flags)** instantiations.

Subsequent patches such as SIMD binary operators involve a total of about 10 operators, and the number of instantiations will also be mutiplied.

How should we simplify this?

**Comment by philnik777** [comment]
@joy2myself IIRC we already reduced that number of ABIs that we test with. I think it would be good enough to test 4-5 U types. That should reduce the time enough to not be problematic anymore.

**Comment by joy2myself** [comment]
> @joy2myself IIRC we already reduced that number of ABIs that we test with. I think it would be good enough to test 4-5 U types. That should reduce the time enough to not be problematic anymore.

@philnik777 Yes, 21 is the number of ABIs after we reduced. I committed a patch to modify the U types to test with https://github.com/llvm/llvm-project/pull/83781. Please review it first. Thank you!

File: libcxx/include/experimental/__simd/simd.h [hunk 1]
```diff
@@ -70,6 +70,19 @@ class simd {
     _Impl::__load(__s_, _Flags::template __apply<simd>(__mem));
   }
 
+  // copy functions
+  template <class _Up, class _Flags>
+  _LIBCPP_HIDE_FROM_ABI enable_if_t<__is_vectorizable_v<_Up> && is_simd_flag_type_v<_Flags>>
+  copy_from(const _Up* __mem, _Flags) {
+    _Impl::__load(__s_, _Flags::template __apply<simd>(__mem));
+  }
```
Comment (philnik777) [inline libcxx/include/experimental/__simd/simd.h:78]:
```suggestion
  template <class _Up, class _Flags, enable_if_t<__is_vectorizable_v<_Up> && is_simd_flag_type_v<_Flags>, int> = 0>
  _LIBCPP_HIDE_FROM_ABI void
  copy_from(const _Up* __mem, _Flags) {
    _Impl::__load(__s_, _Flags::template __apply<simd>(__mem));
  }
```
Same below.

File: libcxx/test/std/experimental/simd/simd.class/simd_copy.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// UNSUPPORTED: c++03, c++11, c++14
+// XFAIL: target=powerpc{{.*}}le-unknown-linux-gnu
```
Comment (philnik777) [inline libcxx/test/std/experimental/simd/simd.class/simd_copy.pass.cpp:10]: Why is this XFAILed?

Comment (joy2myself) [reply libcxx/test/std/experimental/simd/simd.class/simd_copy.pass.cpp]: Removed.

**Comment by joy2myself** [comment]
Do you have any suggestions to solve the AIX errors in buildkite CI?

**Comment by joy2myself** [comment]
I marked XFAIL temporarily for target `powerpc-ibm-aix7.2.5.7` and `powerpc64-ibm-aix7.2.5.7`. The error "Pass-by-value arguments with alignment greater than register width are not supported." seems a target specific limitation, and no related issues have been encountered on other targets. Therefore, this test may require special handling on these two targets. Fix it later.

**Comment by joy2myself** [comment]
@philnik777 Any other comments?

**Review by philnik777 — APPROVED** [review]
Sorry for the long delay. LGTM.

**Pull request merged**